<a href="https://colab.research.google.com/github/evelynda1985/DataScience/blob/master/sklearnMultipleLinearRegressionExercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Multiple Linear Regression with sklearn - Exercise Solution

You are given a real estate dataset.

Real estate is one of those examples that every regression course goes through as it is extremely easy to understand and there is a (almost always) certain causal relationship to be found.

The data is located in the file: 'real_estate_price_size_year.csv'.

You are expected to create a multiple linear regression (similar to the one in the lecture), using the new data.

Apart from that, please:
-  Display the intercept and coefficient(s)
-  Find the R-squared and Adjusted R-squared
-  Compare the R-squared and the Adjusted R-squared
-  Compare the R-squared of this regression and the simple linear regression where only 'size' was used
-  Using the model make a prediction about an apartment with size 750 sq.ft. from 2009
-  Find the univariate (or multivariate if you wish - see the article) p-values of the two variables. What can you say about them?
-  Create a summary table with your findings

In this exercise, the dependent variable is 'price', while the independent variables are 'size' and 'year'.

Good luck!

## Import the relevant libraries

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression

## Load the data

In [4]:
import google.colab
google.colab.files.upload()

Saving real_estate_price_size_year.csv to real_estate_price_size_year.csv


{'real_estate_price_size_year.csv': b'price,size,year\r\n234314.144,643.09,2015\r\n228581.528,656.22,2009\r\n281626.336,487.29,2018\r\n401255.608,1504.75,2015\r\n458674.256,1275.46,2009\r\n245050.28,575.19,2006\r\n265129.064,570.89,2015\r\n175716.48,620.82,2006\r\n331101.344,682.26,2018\r\n218630.608,694.52,2009\r\n279555.096,1060.36,2009\r\n494778.992,1842.51,2009\r\n215472.104,694.52,2015\r\n418753.008,1009.25,2018\r\n444192.008,1300.96,2006\r\n440201.616,1379.72,2006\r\n248337.6,690.54,2018\r\n234178.16,623.94,2006\r\n225451.984,681.07,2006\r\n299416.976,1027.76,2018\r\n268125.08,620.71,2015\r\n171795.24,549.69,2015\r\n412569.472,1207.45,2015\r\n183459.488,518.38,2015\r\n168047.264,525.81,2009\r\n362519.72,1103.3,2018\r\n271793.312,570.89,2018\r\n406852.304,1334.1,2015\r\n297760.44,681.07,2015\r\n368988.432,1496.36,2015\r\n301635.728,1010.33,2006\r\n225452.32,681.07,2006\r\n207742.248,597.9,2009\r\n191486.896,525.81,2015\r\n285223.176,857.54,2018\r\n302000.92,622.97,2018\r\n269225.9

In [10]:
data = pd.read_csv('real_estate_price_size_year.csv')
data

price     size  year
0   234314.144   643.09  2015
1   228581.528   656.22  2009
2   281626.336   487.29  2018
3   401255.608  1504.75  2015
4   458674.256  1275.46  2009
..         ...      ...   ...
95  252460.400   549.80  2009
96  310522.592  1037.44  2009
97  383635.568  1504.75  2006
98  225145.248   648.29  2015
99  274922.856   705.29  2006

[100 rows x 3 columns]

In [33]:
data.describe()

price         size         year
count     100.000000   100.000000   100.000000
mean   292289.470160   853.024200  2012.600000
std     77051.727525   297.941951     4.729021
min    154282.128000   479.750000  2006.000000
25%    234280.148000   643.330000  2009.000000
50%    280590.716000   696.405000  2015.000000
75%    335723.696000  1029.322500  2018.000000
max    500681.128000  1842.510000  2018.000000

## Create the regression

### Declare the dependent and the independent variables

In [11]:
y = data['price']
x = data[['size','year']]

### Regression

In [14]:
reg = LinearRegression()
reg.fit(x,y)

LinearRegression()

### Find the intercept

In [15]:
reg.intercept_

np.float64(-5772267.017463279)

### Find the coefficients

In [16]:
reg.coef_

array([ 227.70085401, 2916.78532684])

### Calculate the R-squared

In [19]:
reg.score(x,y)

0.7764803683276795

### Calculate the Adjusted R-squared

In [21]:
x.shape

(100, 2)

In [22]:
r2 = reg.score(x,y)
n = x.shape[0]
p = x.shape[1]
adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.7718717161282502

### Compare the R-squared and the Adjusted R-squared

Answer...
 R-squared = 0.7764803683276795
 Adjusted R-squared = 0.7718717161282502
 Because is Multiple Linear Regression the most important value is the Adjusted R-squared, and 70% for me not bad
BUT R-square is a lighter bigger than adjusted R-squared, it means we are not penalize for the use of the new variable

### Compare the Adjusted R-squared with the R-squared of the simple linear regression

Answer...
Multiple Linear Regression R-squared = 0.7764803683276795
Simple Linear Regression R-squared = 0.7447391865847586
The difference is not that much, it means there is not much difference with the new variable

### Making predictions

Find the predicted price of an apartment that has a size of 750 sq.ft. from 2009.

In [23]:
reg.predict([[750,2009]])

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([258330.34465995])

### Calculate the univariate p-values of the variables

In [25]:
from sklearn.feature_selection import f_regression
f_regression(x,y)

(array([285.92105192,   0.85525799]), array([8.12763222e-31, 3.57340758e-01]))

In [26]:
p_values = f_regression(x,y)[1]
p_values

array([8.12763222e-31, 3.57340758e-01])

In [28]:
p_values.round(3)

array([0.   , 0.357])

### Create a summary table with your findings

In [29]:
reg_summary = pd.DataFrame(data = x.columns.values, columns=['Features'])

Features
0     size
1     year

In [32]:
reg_summary ['Coefficients'] = reg.coef_
reg_summary ['p-values'] = p_values.round(3)
reg_summary

Features  Coefficients  p-values
0     size    227.700854     0.000
1     year   2916.785327     0.357

Answer... year is not significant for the model, it should be remove